File: midonet-api/src/main/java/org/midonet/api/bgp/rest_api/BgpResource.java
/*
 * Copyright 2014 Midokura SARL
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.midonet.api.bgp.rest_api;

import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

import javax.annotation.security.PermitAll;
import javax.annotation.security.RolesAllowed;
import javax.ws.rs.Consumes;
import javax.ws.rs.DELETE;
import javax.ws.rs.GET;
import javax.ws.rs.POST;
import javax.ws.rs.Path;
import javax.ws.rs.PathParam;
import javax.ws.rs.Produces;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.Response;
import javax.ws.rs.core.SecurityContext;
import javax.ws.rs.core.UriInfo;

import com.google.inject.Inject;
import com.google.inject.assistedinject.Assisted;
import com.google.inject.servlet.RequestScoped;

import org.midonet.api.ResourceUriBuilder;
import org.midonet.api.bgp.Bgp;
import org.midonet.api.bgp.rest_api.AdRouteResource.BgpAdRouteResource;
import org.midonet.api.rest_api.AbstractResource;
import org.midonet.api.rest_api.ResourceFactory;
import org.midonet.api.rest_api.RestApiConfig;
import org.midonet.cluster.DataClient;
import org.midonet.cluster.auth.AuthRole;
import org.midonet.cluster.data.BGP;
import org.midonet.cluster.rest_api.VendorMediaType;
import org.midonet.event.topology.BgpEvent;
import org.midonet.midolman.serialization.SerializationException;
import org.midonet.midolman.state.StateAccessException;

/**
 * Root resource class for bgps.
 */
@RequestScoped
public class BgpResource extends AbstractResource {

    private final static BgpEvent bgpEvent = new BgpEvent();

    private final ResourceFactory factory;

    @Inject
    public BgpResource(RestApiConfig config, UriInfo uriInfo,
                       SecurityContext context, DataClient dataClient,
                       ResourceFactory factory) {
        super(config, uriInfo, context, dataClient, null);
        this.factory = factory;
    }

    /**
     * Handler to deleting BGP.
     *
     * @param id BGP ID from the request.
     * @throws StateAccessException Data access error.
     */
    @DELETE
    @RolesAllowed({ AuthRole.ADMIN, AuthRole.TENANT_ADMIN })
    @Path("{id}")
    public void delete(@PathParam("id") UUID id) throws StateAccessException,
                                                        SerializationException {
        BGP bgpData = dataClient.bgpGet(id);
        if (bgpData == null) {
            return;
        }
        authoriser.tryAuthoriseBgp(id, "delete this BGP");
        dataClient.bgpDelete(id);
        bgpEvent.delete(id, bgpData);
    }

    /**
     * Handler to getting BGP.
     *
     * @param id BGP ID from the request.
     * @throws StateAccessException Data access error.
     * @return A BGP object.
     */
    @GET
    @PermitAll
    @Path("{id}")
    @Produces({ VendorMediaType.APPLICATION_BGP_JSON,
            MediaType.APPLICATION_JSON })
    public Bgp get(@PathParam("id") UUID id) throws StateAccessException,
                                                    SerializationException {

        authoriser.tryAuthoriseBgp(id, "view this BGP.");

        BGP bgpData = dataClient.bgpGet(id);
        if (bgpData == null) {
            throw notFoundException(id, "bgp");
        }

        // Convert to the REST API DTO
        Bgp bgp = new Bgp(bgpData);
        bgp.setBaseUri(getBaseUri());

        return bgp;
    }

    /**
     * Advertising route resource locator for chains.
     *
     * @param id BGP ID from the request.
     * @return BgpAdRouteResource object to handle sub-resource requests.
     */
    @Path("/{id}" + ResourceUriBuilder.AD_ROUTES)
    public BgpAdRouteResource getBgpAdRouteResource(@PathParam("id") UUID id) {
        return factory.getBgpAdRouteResource(id);
    }

    /**
     * Sub-resource class for port's BGP.
     */
    @RequestScoped
    public static class PortBgpResource extends AbstractResource {

        private final UUID portId;

        @Inject
        public PortBgpResource(RestApiConfig config, UriInfo uriInfo,
                               SecurityContext context,
                               DataClient dataClient,
                               @Assisted UUID portId) {
            super(config, uriInfo, context, dataClient, null);
            this.portId = portId;
        }

        /**
         * Handler for creating BGP.
         *
         * @param bgp BGP object.
         * @throws StateAccessException Data access error.
         * @return Response object with 201 status code set if successful.
         */
        @POST
        @RolesAllowed({ AuthRole.ADMIN, AuthRole.TENANT_ADMIN })
        @Consumes({ VendorMediaType.APPLICATION_BGP_JSON,
                    MediaType.APPLICATION_JSON })
        public Response create(Bgp bgp) throws StateAccessException,
                                               SerializationException {

            bgp.setPortId(portId);

            authoriser.tryAuthorisePort(portId, "add BGP to this port.");

            UUID id = dataClient.bgpCreate(bgp.toData());
            bgpEvent.create(id, dataClient.bgpGet(id));
            return Response.created(
                    ResourceUriBuilder.getBgp(getBaseUri(), id))
                    .build();
        }

        /**
         * Handler to getting a list of BGPs.
         *
         * @throws StateAccessException Data access error.
         * @return A list of BGP objects.
         */
        @GET
        @PermitAll
        @Produces({ VendorMediaType.APPLICATION_BGP_COLLECTION_JSON,
                MediaType.APPLICATION_JSON })
        public List<Bgp> list() throws StateAccessException,
                                       SerializationException {

            authoriser.tryAuthorisePort(portId, "view these BGPs.");

            List<BGP> bgpDataList = dataClient.bgpFindByPort(portId);
            List<Bgp> bgpList = new ArrayList<>();
            if (bgpDataList != null) {
                for (BGP bgpData : bgpDataList) {
                    Bgp bgp = new Bgp(bgpData);
                    bgp.setBaseUri(getBaseUri());
                    bgpList.add(bgp);
                }
            }
            return bgpList;
        }
    }
}

File: cluster/midonet-cluster/src/main/scala/org/midonet/cluster/services/rest_api/resources/BgpResource.scala
/*
 * Copyright 2015 Midokura SARL
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.midonet.cluster.services.rest_api.resources

import java.util.{List => JList, UUID}

import javax.ws.rs._
import javax.ws.rs.core.MediaType.APPLICATION_JSON
import javax.ws.rs.core.Response
import javax.ws.rs.core.Response.Status

import scala.collection.JavaConverters._

import com.google.inject.Inject
import com.google.inject.servlet.RequestScoped

import org.midonet.cluster.rest_api.models.{Router, Bgp, RouterPort}
import org.midonet.cluster.services.rest_api.MidonetMediaTypes._
import org.midonet.cluster.services.rest_api.resources.MidonetResource.{Create, Delete, ResourceContext, Update}

@RequestScoped
class BgpResource @Inject()(resContext: ResourceContext)
    extends MidonetResource[Bgp](resContext) {

    @GET
    @Produces(Array(APPLICATION_BGP_JSON,
                    APPLICATION_JSON))
    @Path("{id}")
    override def get(@PathParam("id") id: String,
                     @HeaderParam("Accept") accept: String): Bgp = {
        getResource(classOf[Bgp], id).flatMap(bgp => {
            getResource(classOf[Router], bgp.routerId).map(router => {
                // The legacy v1 API only supports the addition of one BGP peer
                // per port, whereas the v2 architecture only supports one per
                // router. Due to this change, the port identifier is always
                // set to the first router port.
                bgp.localAS = router.localAs
                bgp.portId = router.portIds.get(0)
                bgp
            })
        }).getOrThrow
    }

    @DELETE
    @Path("{id}")
    override def delete(@PathParam("id") id: String): Response = {
        getResource(classOf[Bgp], id).flatMap(bgp => {
            getResource(classOf[Router], bgp.routerId).map(router => {
                router.localAs = Router.NO_AS
                router.bgpNetworkIds.clear()
                multiResource(Seq(Update(router), Delete(classOf[Bgp], id)),
                              MidonetResource.OkNoContentResponse)
            })
        }).getOrThrow
    }

    @Path("{id}/ad_routes")
    def adRoutes(@PathParam("id") id: UUID): BgpAdRouteResource = {
        new BgpAdRouteResource(id, resContext)
    }

}

@RequestScoped
class PortBgpResource @Inject()(portId: UUID, resContext: ResourceContext)
    extends MidonetResource[Bgp](resContext) {

    @GET
    @Produces(Array(APPLICATION_BGP_COLLECTION_JSON,
                    APPLICATION_JSON))
    override def list(@HeaderParam("Accept") accept: String): JList[Bgp] = {
        getResource(classOf[RouterPort], portId).flatMap(port => {
            getResource(classOf[Router], port.routerId).flatMap(router => {
                listResources(classOf[Bgp], router.bgpPeerIds.asScala)
                    .map(list => {
                    list.foreach(bgp => {
                        bgp.localAS = router.localAs
                        bgp.portId = router.portIds.get(0)
                    })
                    list
                })
            })
        }).getOrThrow.asJava
    }

    @POST
    @Consumes(Array(APPLICATION_BGP_JSON,
                    APPLICATION_JSON))
    override def create(bgp: Bgp,
                        @HeaderParam("Content-Type") contentType: String)
    : Response = {
        bgp.setBaseUri(resContext.uriInfo.getBaseUri)
        throwIfViolationsOn(bgp)

        getResource(classOf[RouterPort], portId).flatMap(port => {
            getResource(classOf[Router], port.routerId).map(router => {
                if (router.localAs != Router.NO_AS ||
                    !router.bgpPeerIds.isEmpty) {
                    // Cannot create a BGP peering on a router that already has
                    // an AS number set, or one or more BGP peers.
                    Response.status(Status.CONFLICT).build()
                } else {
                    bgp.create(portId)
                    router.localAs = bgp.localAS
                    multiResource(Seq(Update(port), Create(bgp)),
                                  Response.created(bgp.getUri).build())
                }
            })
        }).getOrThrow
    }

}